src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -71,6 +119,110 @@ public void periodic(){
         else {
             driveHardware.arcadeDrive(arcadeForward, arcadeRotation);
         }
-    } 
 
-}
+        for (int i = 0; i < 4; i++) {
+
+            //SmartDashboard.putBoolean("Drive/modules/"+i+"/status", driveModules[i].getDeadBoolean());
+
+            Rotation2d moduleRotation = new Rotation2d(driveModules[i].getRotationPosition());
+
+            SwerveModuleState optimizedState = SwerveModuleState.optimize(goalModuleStates[i], moduleRotation);
+            double rotationSetpointRadians = optimizedState.angle.getRadians();
+            double speedSetpointMPerS = optimizedState.speedMetersPerSecond;
+
+            double speedRadPerS = speedSetpointMPerS / DriveConstants.wheelRadiusM;
+            double ffVolts = DriveConstants.driveFF.calculate(speedRadPerS);
+            driveModules[i].setDriveVelocity(speedRadPerS, ffVolts);
+
+            double rotationVoltage = rotationPIDs[i].calculate(moduleRotation.getRadians(), rotationSetpointRadians);
+            driveModules[i].setRotationVoltage(rotationVoltage);
```

Comment:
It seems like we should move this into the SwerveModules class. Then we can have a function called controlModule that takes a goal state and does the control for drive power and rotation. DriveSubsystem should orchestrate the modules, and the modules should control themselves.

Reply:
done